src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/format/SelectResultSet.java

```diff
@@ -525,20 +579,66 @@ private void extractData() {
 
             this.rows = populateRows(searchHits);
             this.size = rows.size();
-            this.totalHits = Math.max(size, // size may be greater than totalHits after nested rows be flatten
-                    Optional.ofNullable(searchHits.getTotalHits()).map(th -> th.value)
-                            .orElse(0L));
-
+            this.internalTotalHits = Optional.ofNullable(searchHits.getTotalHits()).map(th -> th.value).orElse(0L);
+            // size may be greater than totalHits after nested rows be flatten
+            this.totalHits = Math.max(size, internalTotalHits);
         } else if (queryResult instanceof Aggregations) {
             Aggregations aggregations = (Aggregations) queryResult;
 
             this.rows = populateRows(aggregations);
             this.size = rows.size();
+            this.internalTotalHits = size;
             // Total hits is not available from Aggregations so 'size' is used
             this.totalHits = size;
         }
     }
 
+    private void populateCursor() {
+        switch(cursor.getType()) {
+            case DEFAULT:
+                populateDefaultCursor((DefaultCursor) cursor);
+            default:
+                return;
+        }
+    }
+
+    private void populateDefaultCursor(DefaultCursor cursor) {
+        /**
+         * Assumption: scrollId, fetchSize, limit already being set in
+         * @see PrettyFormatRestExecutor.buildProtocolForDefaultQuery()
+         */
+
+        Integer limit = cursor.getLimit();
+        long rowsLeft = rowsLeft(cursor.getFetchSize(), cursor.getLimit());
+        if (rowsLeft <= 0) {
```

Comment:
I saw this cursor populate logic and scroll close logic in both here and `CursorResultExecutor`. Is it due to different code path?

Reply:
Yes, this logic is for first page request.